edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,244 @@
+                target.Children.ForEach(child => child.Parent = Option.None<ServiceIdentityTreeNode>());
+
+                // Remove
+                writableSnapshot.Remove(id);
+            }
+
+            // Write back to the atomic field
+            if (!this.nodes.CompareAndSet(snapshot, writableSnapshot.ToImmutableDictionary()))
+            {
+                // There's only ever one thread calling Remove(), so this should never fail
+                throw new InvalidOperationException("Multi-threaded delete detected on ServiceIdentityTree");
+            }
+        }
+
+        public bool Contains(string id)
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.ContainsKey(id);
+        }
+
+        public Option<ServiceIdentity> Get(string id)
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.TryGetValue(id, out ServiceIdentityTreeNode treeNode)
+                    ? Option.Some(treeNode.Identity)
+                    : Option.None<ServiceIdentity>();
+        }
+
+        public IList<string> GetAllIds()
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.Select(kvp => kvp.Value.Identity.Id).ToList();
+        }
+
+        public bool TryGetAuthChain(string id, out string authChain)
+        {
+            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+            authChain = string.Empty;
+
+            // Auth-chain for root
+            if (this.rootDeviceId == id)
+            {
+                authChain = this.rootDeviceId;
+                return true;
+            }
+
+            // Auth-chain for a child somewhere in the tree
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            if (snapshot.TryGetValue(id, out ServiceIdentityTreeNode treeNode))
+            {
+                // Auth chain starts with the originating device
+                authChain = treeNode.Identity.Id;
+
+                // Walk the parents up to the root
+                while (treeNode.Parent.HasValue)
+                {
+                    // Accumulate the auth chain per parent visited in the format:
+                    // "leaf1;edge1;edge2;...;rootEdge"
+                    ServiceIdentityTreeNode parent = treeNode.Parent.Expect(() => new InvalidOperationException());
+                    authChain += ";" + parent.Identity.Id;
+
+                    if (parent.Identity.DeviceId == this.rootDeviceId)
+                    {
+                        // We've reached the root device, we're done
+                        return true;
+                    }
+
+                    // Move on to the next parent
+                    treeNode = parent;
+                }
+
+                // If we walked the parent chain up to a dead-end, i.e. topmost node does not
```

Comment:
Can you pls explain again how dangling trees are possible? If a middle device gets deleted? Should we handle that as part of the remove then?